### PR TITLE
NEW - ESS S/MIME spec schema added

### PR DESCRIPTION
**Pull request by dhensby**
Link to specs: [Rfc 2634](https://tools.ietf.org/rfcmarkup/2634#appendix-A) updated by [Rfc 5035](https://tools.ietf.org/rfcmarkup/5035#appendix-A)

---

This is a first pass at implementing the Enhanced Security Services for S/MIME and its update. I've not used or tested this yet, but it is the leg-work in implementing the spec (hopefully).

I've placed some `@todo` comments into the code around areas that I'm not sure about, specifically:

- There appear to be multiple definitions on explicit size constraints of SEQUENCES and similar structures - is there a way to enforce this? Do we even want to?
- I'm not sure the `ESSPrivacyMark` CHOICE is implemented correctly and I'm not clear on how it can be implemented correctly

It would be great to have another set of eyes over this to check I've not made any silly mistakes and also any feedback on structure would be welcome too (I wasn't sure how to split the files, so just went with what seemed natural).

Further todos:

- [ ] Add some tests

### CODE REVIEW

File: packages/ess/src/content_hints.ts [hunk 1]
```diff
@@ -0,0 +1,22 @@
+import { AsnProp, AsnPropTypes } from '@peculiar/asn1-schema';
+import { ContentType } from '@peculiar/asn1-cms';
+
+/**
+ * ```
+ * ContentHints ::= SEQUENCE {
+ *   contentDescription UTF8String (SIZE (1..MAX)) OPTIONAL,
+ *   contentType ContentType }
+ * ```
+ * @todo - how to implement size?
```
Comment (microshine) [inline packages/ess/src/content_hints.ts:10]: Supporting of `UTF8String (SIZE (1..MAX))` requires `AsnProp` extending

File: packages/ess/src/ess_security_label.ts [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+    @AsnProp({ type: AsnPropTypes.PrintableString })
+    public pString?: string;
+
+    @AsnProp({ type: AsnPropTypes.Utf8String })
+    public utf8String?: string;
+
+    constructor(params: Partial<ESSPrivacyMark> = {}) {
+        // @todo - is this ok? Other `Choice` implementations attempt to discern the type and assign it to the right prop
+        // but that doesn't look like it will be possible with these string types?
+        Object.assign(this, params);
```
Comment (microshine) [inline packages/ess/src/ess_security_label.ts:26]: I think it's ok

File: packages/ess/src/receipts_from.ts [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+ *   -- formerly "allOrNone [0]AllOrNone"
+ *   receiptList [1] SEQUENCE OF GeneralNames }
+ * ```
+ */
+@AsnType({ type: AsnTypeTypes.Choice })
+export class ReceiptsFrom {
+    @AsnProp({ type: AsnPropTypes.Integer, context: 0 })
+    public allOrFirstTier: AllOrFirstTier = AllOrFirstTier.allReceipts;
+
+    @AsnProp({ type: GeneralNames, repeated: "sequence", context: 1 })
```
[the remark this answers is not in the document]
Comment (microshine) [reply packages/ess/src/receipts_from.ts:18]: https://github.com/RanjithaAnkidisamy/Google-Vision-API/blob/993dd9b046ffddb6ea2e366453eb5c90e82560f5/Lib/site-packages/pyasn1_modules/rfc2634.py#L266

[the remark this answers is not in the document]
Comment (dhensby) [reply packages/ess/src/receipts_from.ts]: OK - I've gone and copied where they have implicit tags into this definition 😅

**Comment by microshine** [comment]
@dhensby What is the status of this PR?

**Comment by dhensby** [comment]
@microshine good question! I've been using this in a project and it works as needed, so this is "ready", however I'm aware that there should probably be some tests authored for it!

The outstanding todos feel out of scope and I don't have the time to extend the underlying `AsnProp` decorator at this time, so they can probably just stay as todos?

I'm under a bit of a crunch at the moment, so I won't be able to write tests until next month in all likelihood.

**Comment by microshine** [comment]
I think it's ok. Please publish your PR and I'll merge it and publish

**Comment by dhensby** [comment]
Oh, I did have a problem, which is that I needed a generic `Set` type (https://github.com/PeculiarVentures/asn1-schema/issues/50) and I had to add that to get this working in a meaningful way for my use-case. It would be something good to add in a more generic way to the library.

**Comment by microshine** [comment]
@dhensby Should I merge and publish this schema?

**Comment by dhensby** [comment]
@microshine I'm happy with it, if you are. I've been using it in production for quite some time now :)

**Comment by microshine** [comment]
`@peculiar/asn1-ess@2.1.1` has been published

**Pull request merged**